### PR TITLE
Implement contains function for IP address

### PR DESCRIPTION
**Pull request by ebyhr**
This is a continuation of https://github.com/prestodb/presto/issues/12012. Closed the prestodb's PR.

### CODE REVIEW

**Comment by ebyhr** [comment]
@findepi Sorry, local build succeeded and I didn't notice new check-style. Updated.
https://travis-ci.com/prestosql/presto/jobs/174093734#L2053

File: presto-main/src/main/java/io/prestosql/util/CIDR.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ */
+package io.prestosql.util;
+
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.StringTokenizer;
+
+public abstract class CIDR
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/util/CIDR.java]: Is this class your code or copied from somewhere? If the latter, please include a code comment mentioning the source.  See for ex `io.prestosql.spi.block.IntArrayList`

File: presto-main/src/main/java/io/prestosql/util/CIDR.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        implements Comparable<CIDR>
+{
+    protected InetAddress baseAddress;
+    protected int cidrMask;
+
+    public static CIDR newCIDR(InetAddress baseAddress, int cidrMask)
+            throws UnknownHostException
+    {
+        if (cidrMask < 0) {
+            throw new UnknownHostException("Invalid mask length used: " + cidrMask);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/util/CIDR.java]: I think this is not the right exception to be raised. We were not looking for a host.

**Comment by findepi** [comment]
I am concerned about the amount of object allocations performed at each execution of this new function.
This may render the function unusable for larger data sets. Did you do any stress tests?

File: presto-main/src/main/java/io/prestosql/util/CIDR.java (hunk 1 above)
Comment (ebyhr) [reply presto-main/src/main/java/io/prestosql/util/CIDR.java]: These class (CIDR, CIDR4 & CIDR6) are copied from Netty as David suggested. I'll add comment that.

**Comment by ebyhr** [comment]
I didn't do stress tests. Let me test it and share the result.

File: presto-docs/src/main/sphinx/functions/ipaddress.rst [hunk 3]
```diff
@@ -0,0 +1,12 @@
+====================
+IP Address Functions
+====================
+
+
+.. function:: contains(format, address) -> boolean
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/ipaddress.rst]: I'm thinking we should call the first argument `network`

File: presto-main/src/main/java/io/prestosql/util/CIDR.java (hunk 1 above)
Comment (martint) [inline presto-main/src/main/java/io/prestosql/util/CIDR.java]: All of this seems overkill (CIDR, CIDR4, CIDR6 classes and abstractions). We should be able to implement the `contains` function like this (it might need a few adjustments, but it serves as a starting point):

```java
int separator = cidr.indexOf("/");
if (separator == -1) {
    throw new IllegalArgumentException("Invalid CIDR");
}

int prefixLength = Integer.parseInt(cidr.substring(separator + 1));

if (prefixLength < 0) {
    throw new IllegalArgumentException("Invalid prefix length");
}

if (prefixLength == 0) {
    return true;
}

byte[] base = InetAddresses.forString(cidr.substring(0, separator)).getAddress();
int baseLength = base.length * Byte.SIZE;

if (prefixLength > baseLength) {
    throw new IllegalArgumentException("Prefix length exceeds address length");
}

BigInteger mask = BigInteger.ONE
        .shiftLeft(prefixLength + 1)
        .subtract(BigInteger.ONE)
        .shiftLeft(baseLength - prefixLength);

BigInteger maskedBase = new BigInteger(base)
        .and(mask);

BigInteger maskedAddress = new BigInteger(InetAddresses.forString(ip).getAddress())
        .and(mask);

return maskedBase.equals(maskedAddress);
```

Comment (ebyhr) [reply presto-main/src/main/java/io/prestosql/util/CIDR.java]: In my understanding, we need to write redundant code to cast ipaddress to string. I changed `InetAddresses.forString(ip)` to `InetAddress.getByAddress(address)` using byte[] type.

```java
utf8Slice(InetAddresses.toAddrString(InetAddress.getByAddress(address.getBytes()))).toStringUtf8()
```

Comment (martint) [reply presto-main/src/main/java/io/prestosql/util/CIDR.java]: InetAddress.getByAddress will do a DNS lookup if the address is not an option address, so we don’t want that.

Comment (ebyhr) [reply presto-main/src/main/java/io/prestosql/util/CIDR.java]: InetAddresses.forString uses [InetAddress.getByAddress](https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html#getByAddress-byte:A-) internally and the document says *no reverse name service lookup is performed*, so I think the change is safe.  Am I missing something? 

```java
  public static InetAddress forString(String ipString) {
    byte[] addr = ipStringToBytes(ipString);

    // The argument was malformed, i.e. not an IP string literal.
    if (addr == null) {
      throw formatIllegalArgumentException("'%s' is not an IP string literal.", ipString);
    }

    return bytesToInetAddress(addr);
  }

  private static InetAddress bytesToInetAddress(byte[] addr) {
    try {
      return InetAddress.getByAddress(addr);
    } catch (UnknownHostException e) {
      throw new AssertionError(e);
    }
  }
```

Comment (martint) [reply presto-main/src/main/java/io/prestosql/util/CIDR.java]: Sorry, you’re right. I got it mixed up with getByName

File: presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    private IpAddressFunctions() {}
+
+    @Description("Determines whether given IP address exists in the CIDR")
+    @ScalarFunction
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean contains(@SqlType(StandardTypes.VARCHAR) Slice cidrSlice, @SqlType(StandardTypes.IPADDRESS) Slice address)
+    {
+        String cidr = cidrSlice.toStringUtf8();
+        try {
+            return CIDR.contains(cidr, address.getBytes());
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java]: Let's inline CIDR.contains() here. There's no benefit in that being in a separate class. We can always pull it out later if we ever need to reuse that logic.

File: presto-main/src/main/java/io/prestosql/util/CIDR.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+public class CIDR
+{
+    private CIDR() {}
+
+    public static boolean contains(String cidr, byte[] address)
+            throws UnknownHostException
+    {
+        int separator = cidr.indexOf("/");
+        if (separator == -1) {
+            throw new IllegalArgumentException("Invalid CIDR");
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/util/CIDR.java]: Once you inline it, change this to throw PrestoException directly

File: presto-main/src/main/java/io/prestosql/util/CIDR.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        int prefixLength = Integer.parseInt(cidr.substring(separator + 1));
+
+        if (prefixLength < 0) {
+            throw new IllegalArgumentException("Invalid prefix length");
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/util/CIDR.java]: Same here.

File: presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        catch (IllegalArgumentException | UnknownHostException e) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java]: UnknownHostException is (oddly) thrown by InetAddress.getByAddress when "if IP address is of illegal length". So this would happen if `@SqlType(StandardTypes.IPADDRESS) Slice address` is an invalid IP address (anything other than 4 or 16 bytes). In theory, that should never happen, but let's catch it separately and use a different error message for it.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.testng.annotations.Test;
+
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+
+public class TestIpAddressFunction
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testIpAddressContains()
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: Can we add some tests for:
* edge cases (e.g., the addresses at the edges of the CIDR range)
* combinations of IPv6 CIDR + IPv4 address and IPv4 CIDR + IPv6 addresses. I suspect there are some legal combinations (especially IPv6 CIDR + IPv4 addresses) that are legal and we're not handling correctly.

Comment (ebyhr) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: Added some test cases. By the way, don't we need to care wrong cidr with correct prefix like `x.x.x.x/1`? It will throw IllegalArgumentException at InetAddresses.forString.

```
java.lang.IllegalArgumentException: 'x.x.x.x' is not an IP string literal.
```

Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: Yeah, we should throw PrestoException with INVALID_FUNCTION_ARGUMENT in that case

Comment (ebyhr) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: Let me update it.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        // Prefix length is 0
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '0.0.0.0')", BOOLEAN, true);
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '255.255.255.255')", BOOLEAN, true);
+
+        assertFunction("contains('::ffff:0a00:0001/0', IPADDRESS '0.0.0.0')", BOOLEAN, true);
+        assertFunction("contains('::ffff:0a00:0001/0', IPADDRESS '255.255.255.255')", BOOLEAN, true);
+
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '::ffff:0000:0000')", BOOLEAN, true);
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: I'm not sure these two cases should return `true`. A zero-length prefix is kind of non-sensical, so the result in these cases would likely be defined as a special case in the CIDR rules.

Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: Incidentally, looking at RFC 4632, it would seem a CIDR can't have an arbitrary ip address in the base part. All the bits beyond the prefix might need to be 0. I'll investigate this further.

BTW, here's a table with the possible CIDR notations for IPv4:

```
       notation       addrs/block      # blocks
       --------       -----------     ----------
       n.n.n.n/32               1     4294967296    "host route"
       n.n.n.x/31               2     2147483648    "p2p link"
       n.n.n.x/30               4     1073741824
       n.n.n.x/29               8      536870912
       n.n.n.x/28              16      268435456
       n.n.n.x/27              32      134217728
       n.n.n.x/26              64       67108864
       n.n.n.x/25             128       33554432
       n.n.n.0/24             256       16777216    legacy "Class C"
       n.n.x.0/23             512        8388608
       n.n.x.0/22            1024        4194304
       n.n.x.0/21            2048        2097152
       n.n.x.0/20            4096        1048576
       n.n.x.0/19            8192         524288
       n.n.x.0/18           16384         262144
       n.n.x.0/17           32768         131072
       n.n.0.0/16           65536          65536    legacy "Class B"
       n.x.0.0/15          131072          32768
       n.x.0.0/14          262144          16384
       n.x.0.0/13          524288           8192
       n.x.0.0/12         1048576           4096
       n.x.0.0/11         2097152           2048
       n.x.0.0/10         4194304           1024
       n.x.0.0/9          8388608            512
       n.0.0.0/8         16777216            256    legacy "Class A"
       x.0.0.0/7         33554432            128
       x.0.0.0/6         67108864             64
       x.0.0.0/5        134217728             32
       x.0.0.0/4        268435456             16
       x.0.0.0/3        536870912              8
       x.0.0.0/2       1073741824              4
       x.0.0.0/1       2147483648              2
       0.0.0.0/0       4294967296              1    "default route"
```

Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java]: From the RFC:

```
For example, the legacy "Class B" network 172.16.0.0, with an implied
   network mask of 255.255.0.0, is defined as the prefix 172.16.0.0/16,
   the "/16" indicating that the mask to extract the network portion of
   the prefix is a 32-bit value where the most significant 16 bits are
   ones and the least significant 16 bits are zeros.  Similarly, the
   legacy "Class C" network number 192.168.99.0 is defined as the prefix
   192.168.99.0/24; the most significant 24 bits are ones and the least
   significant 8 bits are zeros.
```

**Comment by ckdarby** [comment]
@ebyhr I couldn't find any docs on the site talking about this and still see the PR open.

Did this die out and not get merged :( ?

**Comment by ebyhr** [comment]
@ckdarby I'm going to fix this PR. It was suspended because there were unclear points about v4 & v6 relationship.

File: presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    private IpAddressFunctions() {}
+
+    @Description("Determines whether given IP address exists in the CIDR")
+    @ScalarFunction
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean contains(@SqlType(StandardTypes.VARCHAR) Slice network, @SqlType(StandardTypes.IPADDRESS) Slice address)
+    {
+        String cidr = network.toStringUtf8();
+
+        int separator = cidr.indexOf("/");
```
Comment (Praveen2112) [inline presto-main/src/main/java/io/prestosql/operator/scalar/IpAddressFunctions.java:41]: Can we `Slice` functions for indexOf, fetching the prefix and for the baseAddress ?

**Comment by pellicano** [comment]
> Let's continue in #9654.
> 
> @pellicano You don't need to keep my commit credit. Please feel free to amend it :)

No worries, I will keep as-is. Thanks for contributing! 🤝
